internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,280 @@
+package jobmanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/edulinq/autograder/internal/lockmanager"
+	"github.com/edulinq/autograder/internal/log"
+	"github.com/edulinq/autograder/internal/util"
+)
+
+type Job[InputType comparable, OutputType any] struct {
+	JobOptions
+
+	JobOutput[InputType, OutputType]
+
+	// A sorted list of items that need work.
+	WorkItems []InputType
+
+	// A function to retrieve existing records.
```

Comment:
You should talk about the semantics here a bit more.
Specifically, an record that can be retrieved is considered "done" and should not be processed.